builder_test.go

```diff
@@ -282,163 +190,426 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
 								{Name: "dep2-next", Version: "v5"},
+								{Name: "dep2-replace", Version: "v6"},
 							},
-						},
-						{
-							Providers: []lifecycle.Buildpack{
-								{ID: "B", Version: "v2"},
+							filepath.Join(appDir, "build-plan-in-B-v2.toml"),
+						)
+
+						testPlan(t,
+							[]lifecycle.Require{
+								{Name: "dep2-next", Version: "v5"},
+								{Name: "dep3", Version: "v1"},
 							},
-							Requires: []lifecycle.Require{
-								{Name: "dep2-replace", Version: "v6"},
+							filepath.Join(appDir, "build-plan-in-C-v1.toml"),
+						)
+					})
+				})
+
+				it("should convert metadata version to top level version for buildpacks with buildpack api 0.2", func() {
+					builder.Plan = lifecycle.BuildPlan{
+						Entries: []lifecycle.BuildPlanEntry{
+							{
+								Providers: []lifecycle.Buildpack{
+									{ID: "B", Version: "v2"},
+								},
+								Requires: []lifecycle.Require{
+									{Name: "dep2", Metadata: map[string]interface{}{"version": "v4"}},
+								},
 							},
 						},
-					},
-				}
+					}
+					_, err := builder.Build()
+					if err != nil {
+						t.Fatalf("Unexpected error:\n%s\n", err)
+					}
+					var bpPlanContents lifecycle.BuildpackPlan
+					_, err = toml.DecodeFile(filepath.Join(appDir, "build-plan-in-B-v2.toml"), &bpPlanContents)
+					if err != nil {
+						t.Fatalf("Unexpected error:\n%s\n", err)
+					}
+					h.AssertEq(t, bpPlanContents.Entries[0].Version, "v4")
+				})
+			})
 
-				mkfile(t,
-					"[[entries]]\n"+
-						`name = "dep1"`+"\n"+
-						`version = "v1"`+"\n"+
-						"[[entries]]\n"+
-						`name = "dep1-replace"`+"\n"+
-						`version = "v7"`+"\n",
-					filepath.Join(appDir, "build-plan-out-A-v1.toml"),
-				)
-				mkfile(t,
-					"[[entries]]\n"+
-						`name = "dep1-next"`+"\n"+
-						`version = "v9"`+"\n"+
-						"[[entries]]\n"+
-						`name = "dep2"`+"\n"+
-						`version = "v4"`+"\n"+
-						"[[entries]]\n"+
-						`name = "dep2-replace"`+"\n"+
-						`version = "v8"`+"\n",
-					filepath.Join(appDir, "build-plan-out-B-v2.toml"),
-				)
-				metadata, err := builder.Build()
-				if err != nil {
-					t.Fatalf("Unexpected error:\n%s\n", err)
-				}
-				if s := cmp.Diff(metadata, &lifecycle.BuildMetadata{
-					Processes: []launch.Process{},
-					Buildpacks: []lifecycle.Buildpack{
-						{ID: "A", Version: "v1", API: "0.3", Homepage: "Buildpack A Homepage"},
-						{ID: "B", Version: "v2", API: "0.2"},
-					},
-					BOM: []lifecycle.BOMEntry{
-						{
-							Require:   lifecycle.Require{Name: "dep1", Version: "v1"},
-							Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
-						},
-						{
-							Require:   lifecycle.Require{Name: "dep1-replace", Version: "v7"},
-							Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
-						},
-						{
-							Require:   lifecycle.Require{Name: "dep1-next", Version: "v9"},
-							Buildpack: lifecycle.Buildpack{ID: "B", Version: "v2"},
+			when("build metadata", func() {
```

Comment:
"build metadata" seems like an overloaded term, maybe build.toml would be better